engine/src/flutter/impeller/display_list/canvas.cc

```diff
@@ -730,6 +743,68 @@ void Canvas::DrawImageRect(const std::shared_ptr<Texture>& image,
   if (!clipped_source) {
     return;
   }
+
+  // Optimization: if the texture has a color filter that is a simple
+  // porter-duff blend or matrix filter, then instead of performing a save layer
+  // we should swap out the shader for the porter duff blend shader and avoid a
+  // saveLayer. This can only be done for imageRects without a strict source
+  // rect, as the porter duff shader does not support this feature. This
+  // optimization is important for flame.
+  if (src_rect_constraint != SourceRectConstraint::kStrict &&  //
+      paint.color_filter &&                                    //
+      paint.image_filter == nullptr &&                         //
+      !paint.mask_blur_descriptor.has_value() &&               //
+      IsPipelineBlendOrMatrixFilter(paint.color_filter)) {
+    if (paint.color_filter->type() == flutter::DlColorFilterType::kBlend) {
```

Comment:
nit: This method is huge, can you split this out into it's own function.

Reply:
Done